// mean-rest-angular/esm5/common.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
var MinNumber = /** @class */ (function () {
    function MinNumber() {
    }
    /**
     * @param {?} control
     * @return {?}
     */
    MinNumber.prototype.validate = /**
     * @param {?} control
     * @return {?}
     */
    function (control) {
        /** @type {?} */
        var num = control.value;
        /** @type {?} */
        var minnum = parseFloat(this.minnumber);
        if (isNaN(minnum))
            return null; //only compare when max is a number
        //only compare when inputs are numbers
        if (typeof num == 'number' && num < minnum)
            return { 'minnumber': "must be no less than " + minnum };
        return null;
    };
    MinNumber.decorators = [
        { type: Directive, args: [{
                    selector: '[minnumber]',
                    providers: [{ provide: NG_VALIDATORS, useExisting: MinNumber, multi: true }]
                },] }
    ];
    MinNumber.propDecorators = {
        minnumber: [{ type: Input, args: ['minnumber',] }]
    };
    return MinNumber;
}());
export { MinNumber };
if (false) {
    /** @type {?} */
    MinNumber.prototype.minnumber;
}
var MaxNumber = /** @class */ (function () {
    function MaxNumber() {
    }
    /**
     * @param {?} control
     * @return {?}
     */
    MaxNumber.prototype.validate = /**
     * @param {?} control
     * @return {?}
     */
    function (control) {
        /** @type {?} */
        var num = control.value;
        /** @type {?} */
        var maxnum = parseFloat(this.maxnumber);
        if (isNaN(maxnum))
            return null; //only compare when max is a number
        //only compare when inputs are numbers
        if (typeof num == 'number' && num > maxnum)
            return { 'maxnumber': "must be no greater than " + maxnum };
        return null;
    };
    MaxNumber.decorators = [
        { type: Directive, args: [{
                    selector: '[maxnumber]',
                    providers: [{ provide: NG_VALIDATORS, useExisting: MaxNumber, multi: true }]
                },] }
    ];
    MaxNumber.propDecorators = {
        maxnumber: [{ type: Input, args: ['maxnumber',] }]
    };
    return MaxNumber;
}());
export { MaxNumber };
if (false) {
    /** @type {?} */
    MaxNumber.prototype.maxnumber;
}
