// mdds-angular-auth/esm5/auth/auth.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject } from '@angular/core';
import { AuthenticationService } from './auth.service';
import { AUTHTICATION_INTERFACES } from './tokens';
var AuthenticationComponent = /** @class */ (function () {
    function AuthenticationComponent(authenticationService, authenticationInterfaces) {
        this.authenticationService = authenticationService;
        this.authenticationInterfaces = authenticationInterfaces;
        this.adminInterface = false;
        this.userIntEnabled = false;
        this.adminIntEnabled = false;
        /** @type {?} */
        var str = authenticationInterfaces.toLowerCase();
        if (str.includes("user") && str.includes("admin")) { //both interface. Check cached.
            this.adminInterface = authenticationService.isAdminInterface();
            this.userIntEnabled = true;
            this.adminIntEnabled = true;
        }
        else if (str.includes("admin")) { //administrator only
            this.authenticationService.setAdminInterface(true);
            this.adminIntEnabled = true;
        }
        else { //users only
            this.authenticationService.setAdminInterface(false);
            this.userIntEnabled = true;
        }
    }
    /**
     * @param {?} adminInterface
     * @return {?}
     */
    AuthenticationComponent.prototype.setAdminInterface = /**
     * @param {?} adminInterface
     * @return {?}
     */
    function (adminInterface) {
        this.adminInterface = adminInterface;
        this.authenticationService.setAdminInterface(this.adminInterface);
    };
    AuthenticationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-auth',
                    template: "<div class=\"container\">\n    <div class=\"row mt-5\">\n        <div class=\"col-md-6 offset-md-3\">\n            <router-outlet></router-outlet>\n        </div>\n    </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    AuthenticationComponent.ctorParameters = function () { return [
        { type: AuthenticationService },
        { type: String, decorators: [{ type: Inject, args: [AUTHTICATION_INTERFACES,] }] }
    ]; };
    return AuthenticationComponent;
}());
export { AuthenticationComponent };
if (false) {
    /** @type {?} */
    AuthenticationComponent.prototype.adminInterface;
    /** @type {?} */
    AuthenticationComponent.prototype.userIntEnabled;
    /** @type {?} */
    AuthenticationComponent.prototype.adminIntEnabled;
    /**
     * @type {?}
     * @private
     */
    AuthenticationComponent.prototype.authenticationService;
    /**
     * @type {?}
     * @private
     */
    AuthenticationComponent.prototype.authenticationInterfaces;
}
