// mean-rest-angular/esm2015/common.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Directive } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
export class MinNumber {
    /**
     * @param {?} control
     * @return {?}
     */
    validate(control) {
        /** @type {?} */
        let num = control.value;
        /** @type {?} */
        let minnum = parseFloat(this.minnumber);
        if (isNaN(minnum))
            return null; //only compare when max is a number
        //only compare when inputs are numbers
        if (typeof num == 'number' && num < minnum)
            return { 'minnumber': "must be no less than " + minnum };
        return null;
    }
}
MinNumber.decorators = [
    { type: Directive, args: [{
                selector: '[minnumber]',
                providers: [{ provide: NG_VALIDATORS, useExisting: MinNumber, multi: true }]
            },] }
];
MinNumber.propDecorators = {
    minnumber: [{ type: Input, args: ['minnumber',] }]
};
if (false) {
    /** @type {?} */
    MinNumber.prototype.minnumber;
}
export class MaxNumber {
    /**
     * @param {?} control
     * @return {?}
     */
    validate(control) {
        /** @type {?} */
        let num = control.value;
        /** @type {?} */
        let maxnum = parseFloat(this.maxnumber);
        if (isNaN(maxnum))
            return null; //only compare when max is a number
        //only compare when inputs are numbers
        if (typeof num == 'number' && num > maxnum)
            return { 'maxnumber': "must be no greater than " + maxnum };
        return null;
    }
}
MaxNumber.decorators = [
    { type: Directive, args: [{
                selector: '[maxnumber]',
                providers: [{ provide: NG_VALIDATORS, useExisting: MaxNumber, multi: true }]
            },] }
];
MaxNumber.propDecorators = {
    maxnumber: [{ type: Input, args: ['maxnumber',] }]
};
if (false) {
    /** @type {?} */
    MaxNumber.prototype.maxnumber;
}
