// mdds-angular-auth/esm2015/auth/auth.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { catchError, map } from 'rxjs/operators';
import { Router, NavigationEnd } from '@angular/router';
import { AUTHTICATION_SERVER_ROOT_URI } from './tokens';
export class AuthenticationService {
    /**
     * @param {?} authServerRootUri
     * @param {?} router
     * @param {?} http
     */
    constructor(authServerRootUri, router, http) {
        this.authServerRootUri = authServerRootUri;
        this.router = router;
        this.http = http;
        this.adminInterface = false;
        this.navigateEndTime = Date.now();
        router.events.subscribe(event => {
            if (event instanceof NavigationEnd) {
                this.previousUrl = this.currentUrl;
                this.navigateEndTime = Date.now();
                if (event.urlAfterRedirects) {
                    this.currentUrl = event.urlAfterRedirects;
                }
                else {
                    this.currentUrl = event.url;
                }
            }
        });
        this.adminInterface = JSON.parse(localStorage.getItem('adminInterface'));
    }
    /**
     * @return {?}
     */
    isAuthorized() {
        /** @type {?} */
        const authRecord = JSON.parse(localStorage.getItem('mdds-auth-record'));
        if (authRecord && authRecord['accessToken']) {
            return true;
        }
        return false;
    }
    /**
     * @private
     * @return {?}
     */
    getLogoutTime() {
        /** @type {?} */
        const authRecord = JSON.parse(localStorage.getItem('mdds-auth-record'));
        if (authRecord && 'logoutTs' in authRecord) {
            return authRecord.logoutTs;
        }
        return 0;
    }
    /**
     * @return {?}
     */
    getAccessToken() {
        /** @type {?} */
        const authRecord = JSON.parse(localStorage.getItem('mdds-auth-record'));
        if (authRecord) {
            return authRecord['accessToken'];
        }
        return null;
    }
    /**
     * @return {?}
     */
    refreshToken() {
        /** @type {?} */
        let authRecord = JSON.parse(localStorage.getItem('mdds-auth-record'));
        if (!authRecord) {
            authRecord = { refreshToken: '', userName: '' };
        }
        /** @type {?} */
        const refreshToken = authRecord['refreshToken'];
        /** @type {?} */
        const userName = authRecord['userName'];
        return this.http.post(this.authServerRootUri + '/refresh', { refreshToken: refreshToken, userName: userName }).pipe(map(this.loggedIn), catchError(error => {
            this.logout();
            return error;
        }));
    }
    /**
     * @param {?} response
     * @return {?}
     */
    refreshShouldHappen(response) {
        return response.status === 401;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    verifyTokenRequest(url) {
        if (url.endsWith(this.authServerRootUri + '/refresh') ||
            url.endsWith(this.authServerRootUri + '/login')) {
            return true;
        }
        return false;
    }
    /**
     * @return {?}
     */
    getInterruptedUrl() {
        return this.interruptedUrl || '/';
    }
    /**
     * @return {?}
     */
    getRoutedFromUrl() {
        return this.routedFromUrl || '/';
    }
    /**
     * @param {?} url
     * @return {?}
     */
    setInterruptedUrl(url) {
        this.interruptedUrl = url;
        /** @type {?} */
        const currentTime = Date.now();
        if (currentTime - this.navigateEndTime > 1000) {
            // Happend > 1 sec. assume it is triggered from current page.
            this.routedFromUrl = this.currentUrl;
        }
        else {
            // page transitioned
            this.routedFromUrl = this.previousUrl;
        }
    }
    /**
     * @return {?}
     */
    getUserName() {
        /** @type {?} */
        let authRecord = JSON.parse(localStorage.getItem('mdds-auth-record'));
        if (!authRecord) {
            authRecord = { userName: '' };
        }
        return authRecord['userName'];
    }
    /**
     * @param {?} userName
     * @param {?} password
     * @return {?}
     */
    login(userName, password) {
        /** @type {?} */
        const authRecord = {
            userName: userName,
            accessToken: '',
            refreshToken: '',
            displayName: ''
        };
        localStorage.setItem('mdds-auth-record', JSON.stringify(authRecord));
        /** @type {?} */
        const options = this.adminInterface ?
            { params: new HttpParams().set('type', 'admin') } : {};
        return this.http.post(this.authServerRootUri + '/login', { username: userName, password: password }, options).pipe(map(this.loggedIn));
    }
    /**
     * @param {?} userInfo
     * @return {?}
     */
    register(userInfo) {
        localStorage.removeItem('mdds-auth-record');
        /** @type {?} */
        const authRecord = {
            userName: userInfo.userName,
            accessToken: '',
            refreshToken: '',
            displayName: userInfo.displayName
        };
        localStorage.setItem('mdds-auth-record', JSON.stringify(authRecord));
        /** @type {?} */
        const options = this.adminInterface ?
            { params: new HttpParams().set('type', 'admin') } : {};
        return this.http.post(this.authServerRootUri + '/register', userInfo, options);
    }
    /**
     * @param {?} user
     * @return {?}
     */
    loggedIn(user) {
        /** @type {?} */
        const authRecord = {
            userName: '',
            accessToken: '',
            refreshToken: '',
            displayName: ''
        };
        if (user && user.accessToken) {
            authRecord['accessToken'] = user.accessToken;
        }
        if (user && user.refreshToken) {
            authRecord['refreshToken'] = user.refreshToken;
        }
        if (user && user.displayName) {
            authRecord['displayName'] = user.displayName;
        }
        if (user && user.userName) {
            authRecord['userName'] = user.userName;
        }
        localStorage.setItem('mdds-auth-record', JSON.stringify(authRecord));
        return user;
    }
    /**
     * @return {?}
     */
    logout() {
        // remove user from local storage to log user out
        /** @type {?} */
        let authRecord = JSON.parse(localStorage.getItem('mdds-auth-record'));
        if (!authRecord) {
            authRecord = {};
        }
        authRecord.logoutTs = Date.now();
        authRecord.accessToken = '';
        authRecord.refreshToken = '';
        localStorage.setItem('mdds-auth-record', JSON.stringify(authRecord));
    }
    /**
     * @param {?} isAdminInterface
     * @return {?}
     */
    setAdminInterface(isAdminInterface) {
        this.adminInterface = isAdminInterface;
        localStorage.setItem('adminInterface', JSON.stringify(isAdminInterface));
    }
    /**
     * @return {?}
     */
    isAdminInterface() {
        return this.adminInterface;
    }
}
AuthenticationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthenticationService.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [AUTHTICATION_SERVER_ROOT_URI,] }] },
    { type: Router },
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.interruptedUrl;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.routedFromUrl;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.previousUrl;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.currentUrl;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.navigateEndTime;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.adminInterface;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.authServerRootUri;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    AuthenticationService.prototype.http;
}
