// mdds-angular-auth/esm2015/auth/auth.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { Router } from '@angular/router';
import { BehaviorSubject, throwError, EMPTY } from 'rxjs';
import { catchError, switchMap, filter, take } from 'rxjs/operators';
import { AuthenticationService } from './auth.service';
import { AUTHTICATION_LOGIN_PAGE_URI } from './tokens';
export class TokenInterceptor {
    /**
     * @param {?} router
     * @param {?} authService
     * @param {?} loginPageUri
     */
    constructor(router, authService, loginPageUri) {
        this.router = router;
        this.authService = authService;
        this.loginPageUri = loginPageUri;
        this.refreshTokenInProgress = false;
        // Refresh Token Subject tracks the current token, or is null if no token is currently
        // available (e.g. refresh pending).
        this.refreshTokenSubject = new BehaviorSubject(null);
    }
    /**
     * @param {?} request
     * @return {?}
     */
    addAuthHeader(request) {
        if (this.authService.isAuthorized()) {
            request = request.clone({
                setHeaders: {
                    Authorization: `Bearer ${this.authService.getAccessToken()}`
                }
            });
        }
        return request;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        return next.handle(this.addAuthHeader(request))
            .pipe(
        /*
        tap((event: HttpEvent<any>) => {
            if (event instanceof HttpResponse) {
              ;
            }
          }, (err: any) => {
            if (err instanceof HttpErrorResponse) {
              if (err.status === 401) {
                this.authService.setInterruptedUrl(this.router.url);
                this.router.navigate([this.loginPageUri]);
              }
            }
        })
        */
        catchError((error) => {
            if (error instanceof HttpErrorResponse && error.status === 401) {
                if (this.refreshTokenInProgress && !this.authService.verifyTokenRequest(request.url)) {
                    // If refreshTokenInProgress is true, we will wait until refreshTokenSubject has a non-null value
                    // – which means the new token is ready and we can retry the request again
                    return this.refreshTokenSubject.pipe(filter(result => result !== null), take(1), switchMap(() => next.handle(this.addAuthHeader(request))));
                }
                else {
                    this.refreshTokenInProgress = true;
                    // Set the refreshTokenSubject to null so that subsequent API calls will wait until the new token has been retrieved
                    this.refreshTokenSubject.next(null);
                    return this.authService.refreshToken().pipe(switchMap((data) => {
                        this.refreshTokenInProgress = false;
                        this.refreshTokenSubject.next(data);
                        return next.handle(this.addAuthHeader(request));
                    }), catchError((err) => {
                        // looks like this part will not hit
                        this.refreshTokenInProgress = false;
                        this.authService.setInterruptedUrl(this.router.url);
                        this.router.navigate([this.loginPageUri]);
                        return EMPTY;
                    }));
                }
            }
            if (this.authService.verifyTokenRequest(request.url)) {
                // refreshToken failed. Go to login page.
                this.refreshTokenInProgress = false;
                this.authService.setInterruptedUrl(this.router.url);
                this.router.navigate([this.loginPageUri]);
            }
            return throwError(error);
        }));
    }
}
TokenInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TokenInterceptor.ctorParameters = () => [
    { type: Router },
    { type: AuthenticationService },
    { type: String, decorators: [{ type: Inject, args: [AUTHTICATION_LOGIN_PAGE_URI,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    TokenInterceptor.prototype.refreshTokenInProgress;
    /**
     * @type {?}
     * @private
     */
    TokenInterceptor.prototype.refreshTokenSubject;
    /**
     * @type {?}
     * @private
     */
    TokenInterceptor.prototype.router;
    /**
     * @type {?}
     * @private
     */
    TokenInterceptor.prototype.authService;
    /**
     * @type {?}
     * @private
     */
    TokenInterceptor.prototype.loginPageUri;
}
