// mean-rest-angular/esm5/base.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { HttpHeaders, HttpParams } from '@angular/common/http';
var ServiceError = /** @class */ (function () {
    function ServiceError() {
    }
    return ServiceError;
}());
export { ServiceError };
if (false) {
    /** @type {?} */
    ServiceError.prototype.status;
    /** @type {?} */
    ServiceError.prototype.clientErrorMsg;
    /** @type {?} */
    ServiceError.prototype.serverError;
}
var BaseService = /** @class */ (function () {
    function BaseService(http, serviceUrl) {
        this.http = http;
        this.serviceUrl = serviceUrl;
        this.storage = {};
    }
    /**
     * @param {?} name
     * @return {?}
     */
    BaseService.prototype.getFromStorage = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.storage[name];
    };
    /**
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    BaseService.prototype.putToStorage = /**
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    function (name, value) {
        this.storage[name] = value;
    };
    /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    BaseService.prototype.errorResponseHandler = /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    function (error) {
        /** @type {?} */
        var err = new ServiceError();
        ;
        if (error.error instanceof ErrorEvent) {
            // A client-side or network error occurred. Handle it accordingly.
            err.status = 0;
            err.clientErrorMsg = error.error.message;
        }
        else {
            // The backend returned an unsuccessful response code.
            // The response body may contain clues as to what went wrong,
            err.status = error.status;
            err.serverError = error.error;
        }
        // return an observable with a user-facing error message
        return throwError(err);
    };
    /**
     * @protected
     * @param {?} detail
     * @return {?}
     */
    BaseService.prototype.formatDetail = /**
     * @protected
     * @param {?} detail
     * @return {?}
     */
    function (detail) {
        return detail;
    };
    /**
     * @protected
     * @param {?} list
     * @return {?}
     */
    BaseService.prototype.formatList = /**
     * @protected
     * @param {?} list
     * @return {?}
     */
    function (list) {
        return list;
    };
    /**
     * @param {?} page
     * @param {?} per_page
     * @param {?} searchContext
     * @return {?}
     */
    BaseService.prototype.getList = /**
     * @param {?} page
     * @param {?} per_page
     * @param {?} searchContext
     * @return {?}
     */
    function (page, per_page, searchContext) {
        /** @type {?} */
        var httpOptions = {
            params: new HttpParams().set('__page', page.toString())
                .set('__per_page', per_page.toString()),
            headers: new HttpHeaders({ 'Accept': 'application/json' }),
        };
        if (!searchContext) {
            return this.http.get(this.serviceUrl, httpOptions)
                .pipe(map(this.formatList), catchError(this.errorResponseHandler));
        }
        httpOptions.params = httpOptions.params.set('action', "Search");
        return this.http.post(this.serviceUrl + "mddsaction/get", searchContext, httpOptions)
            .pipe(catchError(this.errorResponseHandler));
    };
    /**
     * @param {?} page
     * @param {?} per_page
     * @return {?}
     */
    BaseService.prototype.getListWithCondition = /**
     * @param {?} page
     * @param {?} per_page
     * @return {?}
     */
    function (page, per_page) {
        /** @type {?} */
        var httpOptions = {
            params: new HttpParams().set('__page', page.toString())
                .set('__per_page', per_page.toString()),
            headers: new HttpHeaders({ 'Accept': 'application/json' }),
        };
        return this.http.get(this.serviceUrl, httpOptions)
            .pipe(map(this.formatList), catchError(this.errorResponseHandler));
    };
    /**
     * @param {?} id
     * @param {?} action
     * @return {?}
     */
    BaseService.prototype.getDetailForAction = /**
     * @param {?} id
     * @param {?} action
     * @return {?}
     */
    function (id, action) {
        /** @type {?} */
        var httpOptions = {
            headers: new HttpHeaders({ 'Accept': 'application/json' }),
        };
        /** @type {?} */
        var serviceUrl = this.serviceUrl;
        if (action) {
            httpOptions['params'] = new HttpParams().set('action', action);
            if (action == "edit")
                action = "post";
            serviceUrl = serviceUrl + "mddsaction/" + action + "/";
        }
        return this.http.get(serviceUrl + id, httpOptions)
            .pipe(map(this.formatDetail), catchError(this.errorResponseHandler));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    BaseService.prototype.getDetail = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.getDetailForAction(id, null);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    BaseService.prototype.deleteOne = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.delete(this.serviceUrl + id)
            .pipe(catchError(this.errorResponseHandler));
    };
    /**
     * @param {?} ids
     * @return {?}
     */
    BaseService.prototype.deleteManyByIds = /**
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        /** @type {?} */
        var httpOptions = {
            headers: new HttpHeaders({ 'Content-Type': 'application/json' }),
            params: new HttpParams().set('action', "DeleteManyByIds"),
        };
        return this.http.post(this.serviceUrl + "mddsaction/delete", ids, httpOptions)
            .pipe(catchError(this.errorResponseHandler));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    BaseService.prototype.createOne = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var httpOptions = {
            headers: new HttpHeaders({ 'Content-Type': 'application/json' })
        };
        return this.http.put(this.serviceUrl, item, httpOptions)
            .pipe(map(this.formatDetail), catchError(this.errorResponseHandler));
    };
    /**
     * @param {?} id
     * @param {?} item
     * @return {?}
     */
    BaseService.prototype.updateOne = /**
     * @param {?} id
     * @param {?} item
     * @return {?}
     */
    function (id, item) {
        /** @type {?} */
        var httpOptions = {
            headers: new HttpHeaders({ 'Content-Type': 'application/json' })
        };
        return this.http.post(this.serviceUrl + id, item, httpOptions)
            .pipe(catchError(this.errorResponseHandler));
    };
    return BaseService;
}());
export { BaseService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseService.prototype.storage;
    /**
     * @type {?}
     * @protected
     */
    BaseService.prototype.http;
    /**
     * @type {?}
     * @protected
     */
    BaseService.prototype.serviceUrl;
}
