// mdds-angular-auth/esm2015/auth/auth.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject } from '@angular/core';
import { AuthenticationService } from './auth.service';
import { AUTHTICATION_INTERFACES } from './tokens';
export class AuthenticationComponent {
    /**
     * @param {?} authenticationService
     * @param {?} authenticationInterfaces
     */
    constructor(authenticationService, authenticationInterfaces) {
        this.authenticationService = authenticationService;
        this.authenticationInterfaces = authenticationInterfaces;
        this.adminInterface = false;
        this.userIntEnabled = false;
        this.adminIntEnabled = false;
        /** @type {?} */
        let str = authenticationInterfaces.toLowerCase();
        if (str.includes("user") && str.includes("admin")) { //both interface. Check cached.
            this.adminInterface = authenticationService.isAdminInterface();
            this.userIntEnabled = true;
            this.adminIntEnabled = true;
        }
        else if (str.includes("admin")) { //administrator only
            this.authenticationService.setAdminInterface(true);
            this.adminIntEnabled = true;
        }
        else { //users only
            this.authenticationService.setAdminInterface(false);
            this.userIntEnabled = true;
        }
    }
    /**
     * @param {?} adminInterface
     * @return {?}
     */
    setAdminInterface(adminInterface) {
        this.adminInterface = adminInterface;
        this.authenticationService.setAdminInterface(this.adminInterface);
    }
}
AuthenticationComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-auth',
                template: "<div class=\"container\">\n    <div class=\"row mt-5\">\n        <div class=\"col-md-6 offset-md-3\">\n            <router-outlet></router-outlet>\n        </div>\n    </div>\n</div>\n"
            }] }
];
/** @nocollapse */
AuthenticationComponent.ctorParameters = () => [
    { type: AuthenticationService },
    { type: String, decorators: [{ type: Inject, args: [AUTHTICATION_INTERFACES,] }] }
];
if (false) {
    /** @type {?} */
    AuthenticationComponent.prototype.adminInterface;
    /** @type {?} */
    AuthenticationComponent.prototype.userIntEnabled;
    /** @type {?} */
    AuthenticationComponent.prototype.adminIntEnabled;
    /**
     * @type {?}
     * @private
     */
    AuthenticationComponent.prototype.authenticationService;
    /**
     * @type {?}
     * @private
     */
    AuthenticationComponent.prototype.authenticationInterfaces;
}
